sdk/src/event/async_timer/dispatcher.cc

```diff
@@ -9,15 +13,29 @@ namespace async_timer
 {
 std::unique_ptr<event::Timer> Dispatcher::CreateTimer(TimerCallback callback) noexcept
 {
-  (void)callback;
-  return nullptr;
+  return std::unique_ptr<event::Timer>{new (std::nothrow) Timer{callback, *this}};
 }
 
-void Dispatcher::Exit() noexcept {}
+void Dispatcher::Exit() noexcept
+{
+  running_ = false;
+}
 
-void Dispatcher::Run() noexcept {}
+void Dispatcher::Run() noexcept
+{
+  while (running_ && !events_.empty())
+  {
+    auto next_event = events_.begin();
+    std::this_thread::sleep_until(next_event->first);
```

Comment:
How can this sleep be interrupted?

Reply:
I'd recommend a short polling interval over an interrupt mechanism.

I'd rather people use the libevent implementation, but this is meant to be a drop-in replacement that provides the same interface and allows you to use the same codebase but without an event library (in case you only care about synchronous networking and would rather not add an external dependency).

When the dispatcher is built on top of libevent, the background thread makes successive calls to the systems polling function (epoll). Epoll blocks waiting for either a timeout or and event on a file descriptors.

I believe it is possible to artiticially interrupt epolls if you create a dummy file descriptor for it to include in its list and then close the file descriptor forcing an event.

But I prefer, and I think its simpler, to use a polling function with a short timer interval that monitors for termination, closure, and flush events. Basically what's done here
https://github.com/lightstep/lightstep-tracer-cpp/blob/master/src/recorder/stream_recorder/stream_recorder_impl.cpp#L57-L82

Reply:
sleep_until may hang "forever" in gcc-4.8 and vs2017 15.4 when system clock changes, i.e. after NTP time sync.

Reply:
It's uses the steady clock which shouldn't have that problem

Reply:
> it is possible interrupt epoll

According to `man epoll_wait` can be interrupted by a signal handler;

> you create a dummy file descriptor for it to include in its list and then close the file descriptor, forcing an event.

ref: http://cr.yp.to/docs/selfpipe.html